saliency/metrics/pic.py

```diff
@@ -0,0 +1,408 @@
+
+  if not has_color_channel:
+    blurred_img = blurred_img[:, :, 0]
+
+  if issubclass(data_type.type, np.integer):
+    blurred_img = np.round(blurred_img)
+
+  return blurred_img.astype(data_type)
+
+
+def generate_random_mask(image_height: int, image_width: int,
+    fraction=0.01) -> np.ndarray:
+  """Generates a random pixel mask.
+
+    The result of this method can be used as the initial seed to generate
+    a 'fully' blurred image with the help of the create_blurred_image(...)
+    method.
+
+    Args:
+      image_height: the image height for which the mask should be generated.
+      image_width: the image width for which the mask should be generated.
+      fraction: the fraction of the mask pixels that should be set to true.
+        The valid value range is [0.0, 1.0]. Set the value to 0.0 if no
+        information from the original image should be used in the blurred image.
+    Returns:
+      The binary mask with the `fraction` of elements set to True.
+  """
+  mask = np.zeros(shape=[image_height, image_width], dtype=bool)
+  size = mask.size
+  indices = np.random.choice(size, replace=False, size=int(size * fraction))
+  mask[np.unravel_index(indices, mask.shape)] = True
+  return mask
+
+
+def estimate_image_entropy(image: np.ndarray) -> float:
+  """Estimates the amount of information in a given image.
+
+    Args:
+      image: an image, which entropy should be estimated. The dimensions of the
+        array should be [H, W, C] or [H, W] of type uint8.
+    Returns:
+      The estimated amount of information in the image.
+  """
+  buffer = io.BytesIO()
+  pil_image = Image.fromarray(image)
+  pil_image.save(buffer, format='webp', lossless=True, quality=100)
+  buffer.seek(0, os.SEEK_END)
+  length = buffer.tell()
+  buffer.close()
+  return length
+
+
+class ComputePicMetricError(Exception):
+  """An error that can be raised by the compute_pic_metric(...) method.
+
+  See the method description for more information.
+  """
+  pass
+
+
+class ComputePicMetricResult(NamedTuple):
+  """Holds results of compute_pic_metric(...) method."""
+  # x-axis coordinates of PIC curve data points.
+  curve_x: Sequence[float]
+  # y-axis coordinates of PIC curve data points.
+  curve_y: Sequence[float]
+  # A sequence of intermediate blurred images used for PIC computation with
+  # the fully blurred image in front and the original image at the end.
+  blurred_images: Sequence[np.ndarray]
+  # Model predictions for images in the `blurred_images` sequence.
+  predictions: Sequence[float]
+  # Saliency thresholds that were used to generate corresponding
+  # `blurred_images`.
+  thresholds: Sequence[float]
+  # Area under the curve.
+  auc: float
+
+
+def compute_pic_metric(
+    img: np.ndarray,
+    saliency_map: np.ndarray,
+    random_mask: np.ndarray,
+    pred_func: Callable[[np.ndarray], Sequence[float]],
+    saliency_thresholds: Sequence[float],
+    keep_monotonous: bool = False,
```

Comment:
In the original implementation we have all parameters in two classes (one for precompute, one for plotting) which makes it easier to keep track what configuration is used. In this code I see them being distributed around in individual function calls. 

1) Should we discuss on the top module level comment all the hyper-parameters in one place?

2) I couldn't see ignore_wrong_predictions and min_softmax_th options. I think at least ignoring wrong predictions makes a big difference and may be good to have here, as models may not be good at explaining examples that they don't correctly predict. This is probably one of the reasons the colab curves are more upwards trending than saturating like a log curve.

```
class ICParams(object):
  """Parameters for computing PIC/SIC."""
  def __init__(self):
    self.compression_alg = utils.COMPRESSION_WEBP  # webp(default), jpg or png
    self.image_interpolation = 'linear'  # linear or nearest neighbor
    self.background_noise = 'random' 
    self.blend_method = 'bg_ipsal' 
    self.num_steps = 100


class ICPlotParams(object):
  """Parameters for plotting PIC/SIC from computed results."""
  def __init__(self):
    self.information_measure = 'entropy'  # entropy or pixel_prc
    self.curve_interpolation = 'linear'  # previous or linear
    self.force_monotonic = True  # force monotonically increasing curve
    self.clamp_softmax = False  # if true, normalized softmax cannot be >1.0
    self.interpolation_steps = 100  # num points fo interpolate for plotting
    self.ignore_wrong_predictions = True # true is default
    # only pick most confident predictions any result below this softmax will
    # be ignored.
    self.min_softmax_th = 0.9
```

Reply:
@tolga-b, as discussed, I have implemented the changes.